docs/map.rst

```diff
@@ -0,0 +1,629 @@
+Maps
+====
+
+.. highlight:: haskell
+
+Maps, sometimes referred to as dictionaries in other languages, allow you to
+store associations between *unique keys* and *values*. There are three
+implementations provided by the ``containers`` package: `Data.Map.Strict
+<https://hackage.haskell.org/package/containers/docs/Data-Map-Strict.html>`_,
+`Data.Map.Lazy
+<https://hackage.haskell.org/package/containers/docs/Data-Map-Lazy.html>`_, and
+`Data.IntMap
+<https://hackage.haskell.org/package/containers/docs/Data-IntMap.html>`_. You
+almost never want the lazy version so use ``Data.Map.Strict``, or if your keys
+are ``Int`` use ``Data.IntMap``; all of these implementations are *immutable*.
+
+::
+
+    data Map k v = ...
+
+    data IntMap v = ...
+
+.. IMPORTANT::
+   ``Map`` relies on the key type ``k`` having instances of the ``Eq`` and
+   ``Ord`` typeclass for its internal representation. These are already defined
+   for builtin types, and if you are using your own data type you can use the
+   `deriving
+   <https://en.wikibooks.org/wiki/Haskell/Classes_and_types#Deriving>`_
+   mechanism.
+
+
+Short Example
+-------------
+
+The following GHCi session shows some of the basic map functionality::
+
+    import qualified Data.Map.Strict as Map
+
+    let nums = Map.fromList [(1,"one"), (2,"two"), (3,"three")]
+
+    -- Check if we know the english words for 1 and 4
+    Map.member nums 1
+    > True
+
+    Map.member nums 4
+    > False
+
+
+    -- Get the english word for the number 3 and 4.
+    Map.lookup 3 nums
+    > Just "three"
+
+    Map.lookup 4 nums
+    > Nothing
+
+
+    -- Add (4, "four") to our original map.
+    let moreNums = Map.insert 4 "four" nums
+
+    Map.member moreNums 4
+    > True
+
+
+    -- Remove the entry for 1 from our original map.
+    let fewerNums = Map.delete 1 nums
+
+    Map.toAscList fewerNums
+    > [(2,"two"),(3,"three")]
+
+
+    -- Create a new map and combine it with our original map.
+    -- Note that these operations are left-biased.
+    let newNums = Map.fromList [(3,"new three"), (4,"new four")]
+
+    Map.union newNums nums
+    > fromList [(1,"one"),(2,"two"),(3,"new three"),(4,"new four")]
+
+
+Importing Map and IntMap
+------------------------
+
+When using ``Map`` or ``IntMap`` in a Haskell source file you should always use
+a ``qualified`` import because these modules export names that clash with the
+standard Prelude (you can import the type constructor on its own though!).
```

Comment:
In the example below the `Map` type constructor is imported unqualified, `lookup` conflicts with the list version from Prelude so I'm not sure we should recommend that. That leaves `lookup`, which we could recommend importing unqualified, but there's also the `!?` operator which we recommend importing unqualified when we present it.

Reply:
Regardless of whether you import `lookup` qualified, you should *always* `import Prelude hiding (lookup)` if you're using `Map`. Otherwise you may find yourself getting very confusing type errors if you leave off the qualifier by mistake. It looks like you meant to be talking about two different functions in the comment above (not just `lookup`) but things got mangled.

Reply:
Good call, updated section.

The fact that such an easy mistake is a bad type error is really unfortunate, I would really like if GHC had error messages like Elm where it looks at similar functions and known common errors and suggests fixes.